### ISSUE TITLE
[Navigator.NavigationBar] Update to 0.15.0,the previous route navigationBar's button event can be triggered

### ISSUE THREAD

**Issue by everfire130** [facebook/react-native#4559]
if previous page :
![image](https://cloud.githubusercontent.com/assets/11308837/11585630/6a536562-9aa8-11e5-9586-59064bd05b7e.png)

current page:
![image](https://cloud.githubusercontent.com/assets/11308837/11585676/ab63bdea-9aa8-11e5-9352-5aec8ce7169e.png)

![image](https://cloud.githubusercontent.com/assets/11308837/11585708/e5a62d76-9aa8-11e5-9dde-014f0f8515b0.png)
click here ,the previous page button
![image](https://cloud.githubusercontent.com/assets/11308837/11585750/2659c436-9aa9-11e5-86a4-92ae96b14bb4.png)

will be click

**Comment by ashleydw** [comment]
I had the same but fixed it by rendering an empty `<View />` in my `routeMapper` if no button was to be shown

**Comment by everfire130** [comment]
@ashleydw I know that.But how to do this now?
![image](https://cloud.githubusercontent.com/assets/11308837/11605079/72b2f9aa-9b31-11e5-9a47-915331efc0a8.png)

it will be affected no matter use the left 、right or title
and Android‘s effect is different to iOS’s

**Comment by satya164** [comment]
Actually it was happening in earlier versions also. Related #3964

**Comment by adriansprod** [comment]
I have noticed same issue with hidden buttons and NavigatorNavigationBar is creating empty views which block clicks in the nav area. Returning null when no button is required fixed the hidden clicks

**Comment by djk4096** [comment]
I have found this to be a side effect of a change starting in v0.15.0 to how view hit testing is done when opacity=0, in this commit:

 77272d5a9193762b8668c99d54f5bb9a3e423a9e

I am currently working around this by replacing code in NavigatorNavigationBar.js, adding the following prop setting to the View rendered in _getComponent:

```
    rendered = (
      <View
        ref={(ref) => {
          this._components[componentName] = this._components[componentName].set(route, ref);
        }}
        pointerEvents={
          route === this.props.navigator.navigationContext.currentRoute ? 'auto' : 'none'
        }
        style={initialStage[componentName]}>
        {content}
      </View>
    );
```

Also there was just an update to NavigatorNavigationBar.js in master db69004300faf1d850f8127a57859d88205450a7 an hour ago, in this same spot in the code but it does not fix the issue for me.  I am using jumpTo to navigate to routes that are all rendered when the App is initialized.  I also have custom RightButtons per-route.

**Comment by adriansprod** [comment]
Ideally the NavBar should hide the button once faded out rather than build up 0 opacity views.

Another solution is amending NavigatorNavigationBarStylesIOS so the RightButton does not change position when transitioning the view to the left. 

Stages.Left.RightButton sends the button to a left position at 1/3 Screen Width. By removing the 'Left' style value the button fades out in the right and is not blocking the middle of the nav bar at 0 opacity. I can't see a difference visually

**Comment by samuelkraft** [comment]
The fix by @djk4096 works for me!

**Comment by gre** [comment]
@djk4096 I think this was fixed by https://github.com/facebook/react-native/pull/4786 which hopefully should come in next react-native release

**Comment by satya164** [comment]
@gre Thanks. Will test and close.

**Comment by gre** [comment]
Actually I think i've misread what was the issue about, because there is the idea to check the currentRoute in this workaround, maybe the bug is still there actually.

but in @djk4096 workaround, I think we need to use "box-none" instead of "auto" for the reason explained in https://github.com/facebook/react-native/pull/4786

**Comment by samuelkraft** [comment]
The bug is still there after #4786, but as you say will be fixed by using @djk4096 workaround with "box-none".
```js
pointerEvents={
  route === this.props.navigator.navigationContext.currentRoute ? 'box-none' : 'none'
}
```

**Comment by gre** [comment]
yes :)

**Comment by gre** [comment]
thanks

**Comment by djk4096** [comment]
@gre After doing more testing today, I have to say that my workaround is dependent on the NavigationBar being rendered every time a new route is focused (say, by setting state on the main component/parent of the Navigator.)  So, the comparison to the currentRoute was depending on the currentRoute actually changing there...

Otherwise, the left/right buttons are only rendered once, when the nav bar is first rendered.

This means that the workaround would really break things for all cases where re-rendering was not happening... and from further testing, I don't feel that https://github.com/facebook/react-native/pull/4786 alone fixes this case either.

**Comment by djk4096** [comment]
Here's the latest solution that adds the correct 'box-none' to the pointerEvents prop thanks to @gre and @samuelkraft:
```javascript
pointerEvents={
  route === this.props.navigator.navigationContext.currentRoute ? 'box-none' : 'none'
}
```
This sets pointerEvents to work correctly as any component’s view is rendered.

After my previous comment, I realized we need to ensure the pointerEvents are updated whenever the nav bar components don't get re-rendered as a result of a transition.

A way to do this would be for the Navigator to fix up the pointerEvents on nav bar components, once a transition is completed.

A function for this, potentially to be added into NavigatorNavigationBar.js:

```javascript
updatePointerEvents: function(fromIndex, toIndex) {
  var routeFrom = this.props.navState.routeStack[fromIndex];
  var routeTo = this.props.navState.routeStack[toIndex];

  COMPONENT_NAMES.forEach(function (componentName) {
    var map = this._components[componentName];
    var componentFrom = map.get(routeFrom);
    var componentTo = map.get(routeTo);
    componentFrom && componentFrom.setNativeProps({ pointerEvents:'none' });
    componentTo && componentTo.setNativeProps({ pointerEvents:'box-none' });
  }, this);
},
```

It looks like it should be called in Navigator.js within the _completeTransition function, passing the source and destination indexes for the scenes that were just transitioned between.

An acceptable place could be immediately before the line where ```this.state.transitionFromIndex``` gets set to null:

```javascript
this._navBar && this._navBar.updatePointerEvents(this.state.transitionFromIndex, this.state.presentedIndex);
this.state.transitionFromIndex = null;
```
I have tested this a bit, it potentially addresses this additional issue without obvious side effects.

**Comment by ncnlinh** [comment]
Hi, is this problem fixed yet? I'm still getting it in 0.18.1

**Comment by ncnlinh** [comment]
@djk4096 for ping :) thank you

**Comment by djk4096** [comment]
@ncnlinh Hi, yes it is still an issue last I checked.  I have had great success using the patches described in my last message as a workaround to continue to use the Nav Bar.

**Comment by ncnlinh** [comment]
@djk4096 ok thanks, then I guess I will try putting it in my app and see how it goes. :+1: thank you

**Comment by ncnlinh** [comment]
@djk4096 the solution fixes the problem for me. Thanks a lot! Why is this patch not applied and submitted?

**Comment by everfire130** [comment]
The bug had fixed in new version for a long time.I am so sorry that I forget to close the issue and response

**Comment by israelidanny** [comment]
@everfire130 that's not true. I still have this in `0.18.1`. @ncnlinh also reported it.
I don't think it's helpful to close an active, horribly annoying bug.

**Comment by ncnlinh** [comment]
@everfire130 I agree with @israelidanny, It still occurs. Fixing inside the module is only a temporary patch.

**Comment by DaleJefferson** [comment]
@everfire130 This issue is still occurring in 0.18.1 what version has this been fixed in?

**Comment by theosherry** [comment]
I can confirm I'm still having this issue with version 0.20.

**Comment by everfire130** [comment]
@israelidanny @ncnlinh @DaleJefferson @theosherry Right!I haven't tested completely.The issue is still occurring.The solution is mentioned above.

**Comment by chirag04** [comment]
I just encountered this today. Will look into this later today.

**Comment by magrinj** [comment]
Same as @chirag04, I just encountered this bug today with `react-native@0.22.2`.

**Comment by zachguo** [comment]
The bug is still in 0.23.1, @djk4096 can you make a PR?

**Comment by Naoto-Ida** [comment]
Having this issue too on 0.22.2.
I'm doing a `push` from the navigation bar button.

**Comment by lacker** [comment]
As far as I can tell this issue has become a combination of several not-quite-the-same issues which makes it hard to proceed. I am going to close this issue. If people have a bug similar to this one and you can reproduce it on the most recent RN release, feel free to reopen a new issue.
